mongoimport/common.go

```diff
@@ -399,28 +587,130 @@ func validateFields(fields []string) error {
-		// meant to prevent cases where we have fields like "a" and "a.c"
-		for _, latterField := range fieldsCopy[index+1:] {
-			// NOTE: this means we will not support imports that have fields that
-			// include e.g. a, a.b
-			if strings.HasPrefix(latterField, field+".") {
-				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, latterField)
+
+		if index+1 < len(fields) {
+			// For each field except the last, we check that the next field does not
+			// contain the current field as a prefix. Since fields is sorted, this checks
+			// case (3) and case (4).
+			nextField := fields[index+1].field
+
+			if strings.HasPrefix(nextField, field+".") {
+				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, nextField)
+			}
+
+			// This checks case (2).
+			if field == nextField {
+				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, nextField)
+			}
+		}
+
+		if useArrayIndexFields {
+			var nextFieldInfo FieldInfo
+			if index+1 < len(fields) {
+				nextFieldInfo = fields[index+1]
+			}
+			// This checks cases (5), (6), (7), and (8). This is achieved by comparing each field against the
+			// next field in the sorted list.
+			err := validateFieldsUsingArrayIndexes(fieldInfo, nextFieldInfo, index, len(fields))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// validateFieldsUsingArrayIndexes is a helper function for validateFields. It checks for
+// cases (5), (6), (7), and (8). See validateFields for more information on the cases being checked.
+func validateFieldsUsingArrayIndexes(fieldInfo FieldInfo, nextFieldInfo FieldInfo, fieldIndex int, numFields int) (err error) {
+	fieldParts := fieldInfo.parts
+	field := fieldInfo.field
+
+	// First, check each part in the current field against the corresponding part in the next field.
+	for i, part := range fieldParts {
+		if i == 0 {
+			// We don't need to do any checks at the root of the document.
+			// This is because at the root of the document, numerical keys
+			// aren't treated as array indexes.
+			// i.e. the CSV `0\nhello` transforms to BSON `{"0": "hello"}``
+			continue
+		}
+
+		if fieldIndex+1 < numFields && len(nextFieldInfo.parts) >= i+1 {
+			// Check against the next field if we're not already at the last field.
+			// We're checking for cases (5), (6), (7), and (8).
+			// If the next field has fewer field parts than the current field,
+			// the check isn't performed since they can't be inconsistent.
+			// Note, we can infer this because the fields are sorted.
+			err = checkAgainstNextField(fieldInfo, nextFieldInfo, i)
+			if err != nil {
+				return err
 			}
-			// NOTE: this means we will not support imports that have fields like
-			// a, a - since this is invalid in MongoDB
-			if field == latterField {
-				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, latterField)
+		}
+		if num, isNum := isNatNum(fieldParts[i]); isNum && fieldIndex == 0 && num != 0 {
```

Comment:
I don't understand this clause.  The `fieldIndex` test indicates that it only applies to the very first field in the list of sorted fields?!

Reply:
Yes, because `checkAgainstNextField()` actually checks case (6) (indexes must start with a zero) for the next field, not the current field. If we want to check for the current field, we need to look back at the previous field. This means the very first field doesn't get checked. That check occurs here.